File: mac-app/Sources/SkillsRegistryCore/Discover.swift
import Foundation

/// Client for the public SkillNet skill index. Swift mirror of Go
/// `cli/internal/discover` (`discover.Client`, `discover.Response`), so the
/// macOS Discover pane reads the same rows `skills-registry discover --json`
/// prints without shelling out to the CLI binary.
///
/// Two transport facts drive this code, exactly as they do in Go:
///
///   - The endpoint is plain HTTP. SkillNet serves a certificate that does not
///     match the host, so HTTPS cannot be verified and the query travels in
///     plaintext. Only the search terms ever leave the machine.
///   - No credential of any kind is attached. The request is built here rather
///     than through `GitHubAPI`, so there is no path by which the user's
///     GitHub token could reach the index, and the session is ephemeral with
///     cookies and credential storage switched off.
///
/// Every failure (unreachable host, timeout, non-2xx, non-JSON body) throws
/// with no partial response: the caller fails closed and shows an error rather
/// than a half-populated list.
public struct DiscoverClient: Sendable {
    /// The public SkillNet search endpoint. Plain HTTP is intentional.
    public static let defaultBaseURL = "http://api-skillnet.openkg.cn/v1/search"

    /// Overrides `defaultBaseURL`. Tests point it at a fixture, and the app
    /// can point it at a mirror. Same variable the Go client reads.
    public static let baseURLEnv = "SKILLS_DISCOVER_URL"

    /// Labels every response, so a consumer that later gains a second index
    /// can tell the rows apart.
    public static let source = "skillnet"

    public static let defaultLimit = 10
    public static let maxLimit = 50

    /// Hard ceiling on one search, DNS through body read.
    public static let timeout: TimeInterval = 10

    /// Identifies the app to the index. It carries no user identity.
    static let userAgent = "skills-registry-mac-discover"

    /// Caps how much of a response body is accepted, so a hostile or broken
    /// endpoint cannot exhaust memory.
    static let maxBodyBytes = 8 << 20

    public let baseURL: String
    let transport: DiscoverTransporting

    /// `baseURL` defaults to `resolvedBaseURL()`; `transport` defaults to an
    /// ephemeral, credential-free `URLSession`.
    public init(baseURL: String? = nil, transport: DiscoverTransporting? = nil) {
        self.baseURL = baseURL ?? Self.resolvedBaseURL()
        self.transport = transport ?? URLSessionDiscoverTransport()
    }

    /// Resolve the endpoint from `baseURLEnv`, falling back to
    /// `defaultBaseURL`.
    public static func resolvedBaseURL(
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) -> String {
        let override = (environment[baseURLEnv] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return override.isEmpty ? defaultBaseURL : override
    }

    /// Run one search and return the mapped rows.
    public func search(_ query: DiscoverQuery) async throws -> DiscoverResponse {
        let q = try query.normalized()
        let request = try Self.searchRequest(base: baseURL, query: q)
        let (data, response) = try await fetch(request)
        guard (200..<300).contains(response.statusCode) else {
            throw DiscoverError.status(response.statusCode, Self.summarize(data))
        }
        guard let payload = try? JSONDecoder().decode(APIResponse.self, from: data) else {
            throw DiscoverError.notJSON(Self.summarize(data))
        }
        return DiscoverResponse(
            source: Self.source,
            query: q.text,
            mode: q.mode.rawValue,
            results: Self.mapResults(payload.data))
    }

    private func fetch(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response): (Data, HTTPURLResponse)
        do {
            (data, response) = try await transport.get(request)
        } catch let e as DiscoverError {
            throw e
        } catch {
            throw DiscoverError.unreachable(request.url?.absoluteString ?? baseURL,
                                            error.localizedDescription)
        }
        guard data.count <= Self.maxBodyBytes else {
            throw DiscoverError.bodyTooLarge(Self.maxBodyBytes)
        }
        return (data, response)
    }

    /// Build the search request. Query parameters the endpoint already carries
    /// are preserved (a fixture or a mirror may pin its own), and the only
    /// headers set are `Accept` and `User-Agent` — never a credential.
    static func searchRequest(base: String, query: DiscoverQuery) throws -> URLRequest {
        guard var comps = URLComponents(string: base) else {
            throw DiscoverError.invalidEndpoint(base)
        }
        var params = (comps.queryItems ?? []).filter {
            !["q", "mode", "limit", "category"].contains($0.name)
        }
        params.append(URLQueryItem(name: "q", value: query.text))
        params.append(URLQueryItem(name: "mode", value: query.mode.rawValue))
        params.append(URLQueryItem(name: "limit", value: String(query.limit)))
        if !query.category.isEmpty {
            params.append(URLQueryItem(name: "category", value: query.category))
        }
        comps.queryItems = params
        guard let url = comps.url else { throw DiscoverError.invalidEndpoint(base) }

        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        req.httpShouldHandleCookies = false
        req.timeoutInterval = timeout
        return req
    }

    /// Session configuration for the shipped transport: ephemeral, with cookie
    /// and credential storage switched off so nothing the app authenticated
    /// elsewhere can ride along on a plaintext request.
    static func sessionConfiguration() -> URLSessionConfiguration {
        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = timeout
        cfg.timeoutIntervalForResource = timeout
        cfg.httpShouldSetCookies = false
        cfg.httpCookieAcceptPolicy = .never
        cfg.httpCookieStorage = nil
        cfg.urlCredentialStorage = nil
        cfg.requestCachePolicy = .reloadIgnoringLocalCacheData
        return cfg
    }

    /// Project the index payload onto `DiscoverResult`, dropping rows with no
    /// usable identity and collapsing duplicates.
    ///
    /// The index returns clones of the same skill when several repositories
    /// vendor it, so rows are deduplicated on (name, skill_url) with the first
    /// occurrence winning: that keeps the index's own ranking order intact,
    /// and the app never re-sorts by repository popularity.
    static func mapResults(_ data: [APISkill]) -> [DiscoverResult] {
        var out: [DiscoverResult] = []
        var seen = Set<String>()
        out.reserveCapacity(data.count)
        for s in data {
            let row = DiscoverResult(
                name: s.name.trimmed,
                description: s.description.trimmed,
                author: s.author.trimmed,
                category: s.category.trimmed,
                skillURL: s.skillURL.trimmed,
                safety: s.evaluation.safety.level.trimmed,
                completeness: s.evaluation.completeness.level.trimmed,
                executability: s.evaluation.executability.level.trimmed)
            if row.name.isEmpty && row.skillURL.isEmpty { continue }
            let key = row.name + "\u{0}" + row.skillURL
            if seen.contains(key) { continue }
            seen.insert(key)
            out.append(row)
        }
        return out
    }

    /// Render an error body as one short line, so a 500 that returns an HTML
    /// page does not end up inside the UI.
    static func summarize(_ body: Data) -> String {
        let text = String(data: body.prefix(4096), encoding: .utf8) ?? ""
        let collapsed = text.split(whereSeparator: { $0.isWhitespace || $0.isNewline }).joined(separator: " ")
        if collapsed.isEmpty { return "(empty body)" }
        if collapsed.count > 120 { return String(collapsed.prefix(119)) + "…" }
        return collapsed
    }
}

// MARK: - query

/// SkillNet's ranking strategy: literal keyword matching or embedding
/// similarity.
public enum DiscoverMode: String, Sendable, CaseIterable, Identifiable {
    case keyword
    case vector

    public var id: String { rawValue }

    /// Label for a mode toggle.
    public var label: String {
        switch self {
        case .keyword: return "Keyword"
        case .vector: return "Meaning"
        }
    }

    /// One-line explanation of what the mode does.
    public var hint: String {
        switch self {
        case .keyword: return "Literal term matching."
        case .vector: return "Embedding similarity — describe what you want."
        }
    }
}

/// One search request. Defaults and clamping mirror Go `discover.Query`.
public struct DiscoverQuery: Sendable, Equatable {
    public var text: String
    public var mode: DiscoverMode
    public var category: String
    public var limit: Int

    public init(text: String, mode: DiscoverMode = .keyword,
                category: String = "", limit: Int = DiscoverClient.defaultLimit) {
        self.text = text
        self.mode = mode
        self.category = category
        self.limit = limit
    }

    /// Fill in defaults and clamp `limit`. Throws on an empty query, so a
    /// stray keystroke never turns into a request.
    public func normalized() throws -> DiscoverQuery {
        var q = self
        q.text = text.trimmed
        guard !q.text.isEmpty else { throw DiscoverError.emptyQuery }
        q.category = category.trimmed
        if q.limit <= 0 { q.limit = DiscoverClient.defaultLimit }
        if q.limit > DiscoverClient.maxLimit { q.limit = DiscoverClient.maxLimit }
        return q
    }
}

// MARK: - published contract

/// One skill in the index. Field names match the published JSON contract of
/// `skills-registry discover --json`, so the two surfaces cannot drift.
///
/// The three score fields carry SkillNet's `evaluation.<score>.level`
/// (`Good`, `Average`, or `Poor`) and are **empty when the index has no
/// score**. An empty score means unscored and must never render as a pass;
/// `ImportGate.label` is what turns it into text.
public struct DiscoverResult: Sendable, Hashable, Identifiable, Codable {
    public var name: String
    public var description: String
    public var author: String
    public var category: String
    public var skillURL: String
    public var safety: String
    public var completeness: String
    public var executability: String

    public init(name: String, description: String = "", author: String = "",
                category: String = "", skillURL: String = "",
                safety: String = "", completeness: String = "", executability: String = "") {
        self.name = name
        self.description = description
        self.author = author
        self.category = category
        self.skillURL = skillURL
        self.safety = safety
        self.completeness = completeness
        self.executability = executability
    }

    /// Stable identity for a SwiftUI list. The index can return two rows with
    /// the same name from different repositories, so the URL is part of it.
    public var id: String { skillURL + "\u{0}" + name }

    /// The index's grades for this row, in the shape the import gate reads.
    public var scores: ImportScores {
        ImportScores(safety: safety, completeness: completeness, executability: executability)
    }

    enum CodingKeys: String, CodingKey {
        case name, description, author, category
        case skillURL = "skill_url"
        case safety, completeness, executability
    }
}

/// The published search payload. Mirrors Go `discover.Response`.
public struct DiscoverResponse: Sendable, Codable, Equatable {
    public var source: String
    public var query: String
    public var mode: String
    public var results: [DiscoverResult]

    public init(source: String, query: String, mode: String, results: [DiscoverResult]) {
        self.source = source
        self.query = query
        self.mode = mode
        self.results = results
    }
}

// MARK: - errors

/// A search failure. Distinct cases so the UI can say what went wrong without
/// parsing a message.
public enum DiscoverError: Error, LocalizedError, Equatable {
    case emptyQuery
    case invalidEndpoint(String)
    case unreachable(String, String)
    case status(Int, String)
    case notJSON(String)
    case bodyTooLarge(Int)

    /// Appended to every failure surfaced to the user. The index is a
    /// convenience, not a dependency: a user who already has a GitHub URL can
    /// still import it from the Add pane.
    public static let fallbackHint =
        "The skill index is optional — you can still import any skill from a GitHub URL in Add."

    public var errorDescription: String? {
        switch self {
        case .emptyQuery:
            return "Enter something to search for."
        case .invalidEndpoint(let base):
            return "Invalid skill index URL \(base)."
        case .unreachable(_, let why):
            return "Couldn't reach the skill index: \(why)"
        case .status(let code, let body):
            return "The skill index returned HTTP \(code): \(body)"
        case .notJSON(let body):
            return "The skill index returned a non-JSON response: \(body)"
        case .bodyTooLarge(let limit):
            return "The skill index returned more than \(limit) bytes."
        }
    }
}

// MARK: - transport

/// One GET against the index. `DiscoverClient` depends on this rather than on
/// `URLSession` directly so tests can inspect the exact request and script
/// failures without touching the network.
public protocol DiscoverTransporting: Sendable {
    func get(_ request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

struct URLSessionDiscoverTransport: DiscoverTransporting {
    let session: URLSession

    init(session: URLSession? = nil) {
        self.session = session ?? URLSession(configuration: DiscoverClient.sessionConfiguration())
    }

    func get(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw DiscoverError.unreachable(request.url?.absoluteString ?? "", "no HTTP response")
        }
        return (data, http)
    }
}

// MARK: - wire models

/// The subset of SkillNet's payload this client reads. The index's own extra
/// fields (repository star count, per-score prose, cost and maintainability
/// grades) are deliberately dropped, matching the Go client: star counts
/// belong to the host repository rather than the individual skill.
struct APIResponse: Decodable {
    var data: [APISkill] = []

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        // A payload with no `data` key is an empty result set, matching Go's
        // `json.Unmarshal` into a nil slice. A `data` of the wrong type still
        // throws, so a 2xx that does not honor the contract fails closed.
        data = try c.decodeIfPresent([APISkill].self, forKey: .data) ?? []
    }

    enum CodingKeys: String, CodingKey { case data }
}

struct APISkill: Decodable {
    var name = ""
    var description = ""
    var author = ""
    var category = ""
    var skillURL = ""
    var evaluation = APIEvaluation()

    enum CodingKeys: String, CodingKey {
        case name = "skill_name"
        case description = "skill_description"
        case author, category
        case skillURL = "skill_url"
        case evaluation
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ""
        description = try c.decodeIfPresent(String.self, forKey: .description) ?? ""
        author = try c.decodeIfPresent(String.self, forKey: .author) ?? ""
        category = try c.decodeIfPresent(String.self, forKey: .category) ?? ""
        skillURL = try c.decodeIfPresent(String.self, forKey: .skillURL) ?? ""
        evaluation = try c.decodeIfPresent(APIEvaluation.self, forKey: .evaluation) ?? APIEvaluation()
    }

    init(name: String, skillURL: String) {
        self.name = name
        self.skillURL = skillURL
    }
}

struct APIEvaluation: Decodable {
    var safety = APIScore()
    var completeness = APIScore()
    var executability = APIScore()

    init() {}

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        safety = try c.decodeIfPresent(APIScore.self, forKey: .safety) ?? APIScore()
        completeness = try c.decodeIfPresent(APIScore.self, forKey: .completeness) ?? APIScore()
        executability = try c.decodeIfPresent(APIScore.self, forKey: .executability) ?? APIScore()
    }

    enum CodingKeys: String, CodingKey { case safety, completeness, executability }
}

/// One grade. A missing `level` stays empty: an unscored skill must not be
/// promoted to a passing grade.
struct APIScore: Decodable {
    var level = ""

    init() {}

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        level = try c.decodeIfPresent(String.self, forKey: .level) ?? ""
    }

    enum CodingKeys: String, CodingKey { case level }
}

extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

File: cli/internal/discover/discover.go
// Package discover searches the public SkillNet skill index and maps its
// responses onto this project's own result shape.
//
// The index is a third-party read-only service reached with one
// unauthenticated GET. Two transport facts drive the code here:
//
//   - The endpoint is plain HTTP. SkillNet serves a certificate that does not
//     match the host, so HTTPS cannot be verified and queries travel in
//     plaintext. Only the search terms are ever sent.
//   - No credential of any kind is attached. The request carries no GitHub
//     token, no `gh` auth header, no cookie, and no registry content, so a
//     plaintext hop leaks nothing beyond the query itself.
//
// Search maps SkillNet's payload onto Response, which is the published
// contract the CLI's `--json` output, the TUI, and the macOS app all consume.
// SkillNet's own fields (the repository star count, per-score prose, cost and
// maintainability grades) are deliberately dropped: star counts belong to the
// host repository rather than the individual skill, and a stable narrow shape
// is what downstream surfaces need.
package discover

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path"
	"strconv"
	"strings"
	"time"

	"github.com/nikships/skills-registry/cli/internal/registry"
)

const (
	// DefaultBaseURL is the public SkillNet search endpoint. Plain HTTP is
	// intentional: see the package comment.
	DefaultBaseURL = "http://api-skillnet.openkg.cn/v1/search"

	// BaseURLEnv overrides DefaultBaseURL. Tests point it at an httptest
	// server, and the macOS app can point it at a mirror.
	BaseURLEnv = "SKILLS_DISCOVER_URL"

	// SourceSkillNet labels every Response so a consumer that later gains a
	// second index can tell the rows apart.
	SourceSkillNet = "skillnet"

	// DefaultLimit and MaxLimit bound how many rows one search asks for. The
	// index holds tens of thousands of skills, so an unbounded limit would
	// return an unreadable page.
	DefaultLimit = 10
	MaxLimit     = 50

	// Timeout is the hard ceiling on one search, DNS through body read.
	Timeout = 10 * time.Second

	// userAgent identifies this CLI to the index. It carries no user
	// identity.
	userAgent = "skills-registry-discover"

	// maxBodyBytes caps how much of a response body is read, so a hostile or
	// broken endpoint cannot exhaust memory.
	maxBodyBytes = 8 << 20
)

// Mode selects SkillNet's ranking strategy: literal keyword matching or
// embedding similarity.
type Mode string

const (
	ModeKeyword Mode = "keyword"
	ModeVector  Mode = "vector"
)

// ParseMode validates a user-supplied --mode value.
func ParseMode(s string) (Mode, error) {
	switch Mode(strings.ToLower(strings.TrimSpace(s))) {
	case ModeKeyword, "":
		return ModeKeyword, nil
	case ModeVector:
		return ModeVector, nil
	default:
		return "", fmt.Errorf("unknown mode %q (want %s or %s)", s, ModeKeyword, ModeVector)
	}
}

// Query is one search request.
type Query struct {
	// Text is the search string. Required.
	Text string
	// Mode defaults to ModeKeyword when empty.
	Mode Mode
	// Category, when set, restricts results to that SkillNet category.
	Category string
	// Limit is the requested row count, defaulted to DefaultLimit and capped
	// at MaxLimit.
	Limit int
}

// normalize fills in defaults and clamps Limit. Callers get the effective
// query back so the JSON payload reports what was actually asked for.
func (q Query) normalize() (Query, error) {
	q.Text = strings.TrimSpace(q.Text)
	if q.Text == "" {
		return Query{}, errors.New("search query is empty")
	}
	mode, err := ParseMode(string(q.Mode))
	if err != nil {
		return Query{}, err
	}
	q.Mode = mode
	q.Category = strings.TrimSpace(q.Category)
	switch {
	case q.Limit <= 0:
		q.Limit = DefaultLimit
	case q.Limit > MaxLimit:
		q.Limit = MaxLimit
	}
	return q, nil
}

// Result is one skill in the index, in this project's shape.
//
// The three score fields carry SkillNet's `evaluation.<score>.level`
// (`Good`, `Average`, or `Poor`) and are empty when the index has no score.
// An empty score means unscored; it must never be rendered as a pass.
type Result struct {
	Name          string `json:"name"`
	Description   string `json:"description"`
	Author        string `json:"author"`
	Category      string `json:"category"`
	SkillURL      string `json:"skill_url"`
	Safety        string `json:"safety"`
	Completeness  string `json:"completeness"`
	Executability string `json:"executability"`
}

// Response is the published search payload. Results is always non-nil so it
// encodes as `[]` rather than `null`.
type Response struct {
	Source  string   `json:"source"`
	Query   string   `json:"query"`
	Mode    string   `json:"mode"`
	Results []Result `json:"results"`
}

// Client searches one index endpoint.
type Client struct {
	// BaseURL is the search endpoint. Empty means BaseURL().
	BaseURL string
	// HTTPClient overrides the default client. Tests inject one with a short
	// timeout; production callers leave it nil.
	HTTPClient *http.Client
}

// New returns a Client bound to the configured endpoint.
func New() *Client {
	return &Client{BaseURL: BaseURL()}
}

// BaseURL resolves the endpoint from BaseURLEnv, falling back to
// DefaultBaseURL.
func BaseURL() string {
	if v := strings.TrimSpace(os.Getenv(BaseURLEnv)); v != "" {
		return v
	}
	return DefaultBaseURL
}

// Search runs one query and returns the mapped results.
//
// Every failure mode (unreachable host, timeout, non-2xx status, non-JSON
// body) returns an error with no partial Response: the caller is expected to
// fail closed rather than present a half-populated list.
func (c *Client) Search(ctx context.Context, q Query) (Response, error) {
	q, err := q.normalize()
	if err != nil {
		return Response{}, err
	}
	endpoint, err := searchURL(c.endpoint(), q)
	if err != nil {
		return Response{}, err
	}
	body, err := c.get(ctx, endpoint)
	if err != nil {
		return Response{}, err
	}
	var payload apiResponse
	if err := json.Unmarshal(body, &payload); err != nil {
		return Response{}, fmt.Errorf("skill index returned a non-JSON response: %w", err)
	}
	return Response{
		Source:  SourceSkillNet,
		Query:   q.Text,
		Mode:    string(q.Mode),
		Results: mapResults(payload.Data),
	}, nil
}

func (c *Client) endpoint() string {
	if strings.TrimSpace(c.BaseURL) != "" {
		return c.BaseURL
	}
	return BaseURL()
}

// get performs the search request. It builds the request itself rather than
// reusing any shared GitHub transport, so there is no path by which an
// Authorization header could reach the index.
func (c *Client) get(ctx context.Context, endpoint string) ([]byte, error) {
	client := c.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: Timeout}
	}
	ctx, cancel := context.WithTimeout(ctx, Timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", userAgent)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("reach the skill index at %s: %w", endpoint, err)
	}
	defer func() { _ = resp.Body.Close() }()

	body, err := io.ReadAll(io.LimitReader(resp.Body, maxBodyBytes))
	if err != nil {
		return nil, fmt.Errorf("read the skill index response: %w", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return nil, fmt.Errorf("skill index returned HTTP %d: %s", resp.StatusCode, summarize(body))
	}
	return body, nil
}

// searchURL appends the query parameters to the endpoint, preserving any the
// endpoint already carries (an httptest server or a mirror may pin its own).
func searchURL(base string, q Query) (string, error) {
	u, err := url.Parse(base)
	if err != nil {
		return "", fmt.Errorf("invalid skill index URL %q: %w", base, err)
	}
	params := u.Query()
	params.Set("q", q.Text)
	params.Set("mode", string(q.Mode))
	params.Set("limit", strconv.Itoa(q.Limit))
	if q.Category != "" {
		params.Set("category", q.Category)
	}
	u.RawQuery = params.Encode()
	return u.String(), nil
}

// summarize renders an error body as a single short line, so a 500 that
// returns an HTML page does not dump the page into the terminal.
func summarize(body []byte) string {
	s := strings.Join(strings.Fields(string(body)), " ")
	if s == "" {
		return "(empty body)"
	}
	if r := []rune(s); len(r) > 120 {
		return string(r[:119]) + "…"
	}
	return s
}

// mapResults projects the index payload onto Result, dropping rows with no
// usable identity and collapsing duplicates.
//
// The index returns clones of the same skill when several repositories vendor
// it, so rows are deduplicated on (name, skill_url) with the first occurrence
// winning: that keeps the index's own ranking order intact.
func mapResults(data []apiSkill) []Result {
	out := make([]Result, 0, len(data))
	seen := make(map[[2]string]bool, len(data))
	for _, s := range data {
		r := Result{
			Name:          strings.TrimSpace(s.Name),
			Description:   strings.TrimSpace(s.Description),
			Author:        strings.TrimSpace(s.Author),
			Category:      strings.TrimSpace(s.Category),
			SkillURL:      strings.TrimSpace(s.SkillURL),
			Safety:        s.Evaluation.Safety.value(),
			Completeness:  s.Evaluation.Completeness.value(),
			Executability: s.Evaluation.Executability.value(),
		}
		if r.Name == "" && r.SkillURL == "" {
			continue
		}
		key := [2]string{r.Name, r.SkillURL}
		if seen[key] {
			continue
		}
		seen[key] = true
		out = append(out, r)
	}
	return out
}

// SkillKey reduces a skill URL to an identity that survives a revision
// change: "owner/repo/path", lowercased. The index links a skill at whatever
// commit it last indexed, so the SHA in a URL the user pasted and the SHA in
// the index's own row routinely differ while naming the same folder. The bool
// is false for anything that is not a github.com folder URL.
func SkillKey(rawURL string) (string, bool) {
	target, ok := registry.ParseGitHubURL(rawURL)
	if !ok || !target.IsFolder() {
		return "", false
	}
	return strings.ToLower(target.FullName() + "/" + strings.Trim(target.Path, "/")), true
}

// Lookup finds the index's row for one skill folder URL, so a caller that
// already has a URL can read the grades the index assigned to it.
//
// The index has no lookup-by-URL endpoint, so this searches for the folder's
// own name and keeps the row whose SkillKey matches. A miss is not an error:
// the index simply has no row, which means unscored. The caller must treat
// unscored as unvetted rather than as a pass.
func (c *Client) Lookup(ctx context.Context, skillURL string) (Result, bool, error) {
	key, ok := SkillKey(skillURL)
	if !ok {
		return Result{}, false, nil
	}
	name := path.Base(key)
	if name == "." || name == "/" || name == "" {
		return Result{}, false, nil
	}
	resp, err := c.Search(ctx, Query{Text: name, Limit: MaxLimit})
	if err != nil {
		return Result{}, false, err
	}
	for _, r := range resp.Results {
		if got, ok := SkillKey(r.SkillURL); ok && got == key {
			return r, true, nil
		}
	}
	return Result{}, false, nil
}

// apiResponse is the subset of SkillNet's payload this package reads.
type apiResponse struct {
	Data []apiSkill `json:"data"`
}

type apiSkill struct {
	Name        string `json:"skill_name"`
	Description string `json:"skill_description"`
	Author      string `json:"author"`
	Category    string `json:"category"`
	SkillURL    string `json:"skill_url"`
	Evaluation  struct {
		Safety        apiScore `json:"safety"`
		Completeness  apiScore `json:"completeness"`
		Executability apiScore `json:"executability"`
	} `json:"evaluation"`
}

type apiScore struct {
	Level string `json:"level"`
}

// value returns the score's level verbatim. A missing level stays empty: an
// unscored skill must not be promoted to a passing grade.
func (s apiScore) value() string { return strings.TrimSpace(s.Level) }
